src/perf_mainloop.cc

```diff
@@ -0,0 +1,290 @@
+  for (int i = 0; i < *pfd_len; ++i) {
+    // NOTE: if fd is negative, it will be ignored
+    pfd[i].fd = pes[i].fd;
+    pfd[i].events = POLLIN | POLLERR | POLLHUP;
+  }
+}
+
+static DDRes signalfd_setup(pollfd *pfd) {
+  sigset_t mask;
+
+  sigemptyset(&mask);
+  sigaddset(&mask, SIGINT);
+  sigaddset(&mask, SIGTERM);
+
+  // no need to block signal, since we inherited sigprocmask from parent
+  int sfd = signalfd(-1, &mask, 0);
+  DDRES_CHECK_ERRNO(sfd, DD_WHAT_WORKERLOOP_INIT, "Could not set signalfd");
+
+  pfd->fd = sfd;
+  pfd->events = POLLIN;
+  return {};
+}
+
+static DDRes worker_loop(DDProfContext *ctx, const WorkerAttr *attr,
+                         bool *restart_worker) {
+
+  // Setup poll() to watch perf_event file descriptors
+  int pe_len = ctx->worker_ctx.pevent_hdr.size;
+  // one extra slot in pfd to accomodate for signal fd
+  struct pollfd pfd[MAX_NB_WATCHERS + 1];
+  int pfd_len = 0;
+  pollfd_setup(&ctx->worker_ctx.pevent_hdr, pfd, &pfd_len);
+
+  DDRES_CHECK_FWD(signalfd_setup(&pfd[pfd_len]));
+  int signal_pos = pfd_len++;
+
+  // Perform user-provided initialization
+  defer { attr->finish_fun(ctx); };
+  DDRES_CHECK_FWD(attr->init_fun(ctx));
+
+  // Worker poll loop
+  while (true) {
+    int n = poll(pfd, pfd_len, PSAMPLE_DEFAULT_WAKEUP_MS);
+
+    // If there was an issue, return and let the caller check errno
+    if (-1 == n && errno == EINTR) {
+      continue;
+    }
+    DDRES_CHECK_ERRNO(n, DD_WHAT_POLLERROR, "poll failed");
+
+    if (pfd[signal_pos].revents & POLLIN) {
+      LG_NFO("Received termination signal");
+      break;
+    }
+
+    // Convenience structs
+    PEvent *pes = ctx->worker_ctx.pevent_hdr.pes;
+
+    // If one of the perf_event_open() feeds was closed by the kernel, shut down
+    // profiling
+    if (std::any_of(pfd, pfd + pe_len,
+                    [&](pollfd &x) { return x.revents & POLLHUP; })) {
+      break;
+    }
+
+    // While there are events to process, iterate through them
+    // while limiting time spent in loop to at most PSAMPLE_DEFAULT_WAKEUP_MS
+    int64_t loop_start_ns = now_nanos();
+    int64_t now_ns = loop_start_ns;
+    bool events;
+    do {
+      events = false;
+      for (int i = 0; i < pe_len; ++i) {
+        // Memory-ordering safe access of ringbuffer elements
+        RingBuffer *rb = &pes[i].rb;
+        uint64_t head = rb->region->data_head;
+        rmb();
+        uint64_t tail = rb->region->data_tail;
+        if (head == tail)
+          continue;
+        events = true;
+
+        // Attempt to dispatch the event
+        struct perf_event_header *hdr = rb_seek(rb, tail);
+        DDRes res = ddprof_worker_process_event(hdr, pes[i].pos, ctx);
+
+        // We've processed the current event, so we can advance the ringbuffer
+        rb->region->data_tail += hdr->size;
+
+        // Check for processing error
+        if (IsDDResNotOK(res)) {
+          return res;
+        }
+      }
+
+      now_ns = now_nanos();
```

Comment:
`now_nanos()` is going to issue a call to `gettimeofday()`, which in many contemporary cloud configurations will skip vDSO, emitting a syscall.  I would like this to be a problem in the future, but it isn't a problem now.

For samples this is (unfortunately!!!!!!) not a problem because libdwfl already calls many aggressive syscalls all the time 🙃 .

For events which are not samples, this will generally change the sample from a pure in-process computation to a syscall.  This might have some disruptions to our event processing queue, but I think this is probably better for now.

Reply:
Followup comment:  my presumption that we weren't already calling `gettimeofday()` in this loop was wrong, and my comments are wrong.  Keeping them for the sake of exposition, but they are wrong.